parquet/src/encryption/encrypt.rs

```diff
@@ -374,6 +395,33 @@ pub(crate) fn encrypt_object<T: TSerializable, W: Write>(
     Ok(())
 }
 
+pub(crate) fn sign_and_write_object<T: TSerializable, W: Write>(
+    object: &T,
+    encryptor: &mut Box<dyn BlockEncryptor>,
+    sink: &mut W,
+    module_aad: &[u8],
+) -> Result<()> {
+    let mut buffer: Vec<u8> = vec![];
+    {
+        let mut protocol = TCompactOutputProtocol::new(&mut buffer);
+        object.write_to_out_protocol(&mut protocol)?;
+    }
+    sink.write_all(&buffer)?;
+    buffer = encryptor.encrypt(buffer.as_ref(), module_aad)?;
+
+    let ciphertext_length : u32 = buffer.len()
+        .try_into()
+        .map_err(|err| general_err!("Plaintext data too long. {:?}", err))?;
+
+    // Format is: [ciphertext size, nonce, ciphertext, authentication tag]
+    let nonce = buffer[SIZE_LEN..SIZE_LEN + NONCE_LEN].to_vec();
+    let tag = buffer[(ciphertext_length - TAG_LEN as u32) as usize..].to_vec();
+    sink.write_all(&nonce)?;
+    sink.write_all(&tag)?;
```

Comment:
I don't think you need the u32 `ciphertext_length` above, this can be simplified. Plus you don't need to copy the nonce and tag to new vectors:

```suggestion
    let nonce = &buffer[SIZE_LEN..SIZE_LEN + NONCE_LEN];
    let tag = &buffer[(buffer.len() - TAG_LEN)..];
    sink.write_all(nonce)?;
    sink.write_all(tag)?;
```